dashboard/test/ui/runner.rb

```diff
@@ -38,187 +39,246 @@
-$options.hourofcode_domain = 'test.hourofcode.com'
-$options.local = nil
-$options.html = nil
-$options.out = nil
-$options.maximize = nil
-$options.auto_retry = false
-$options.magic_retry = false
-$options.parallel_limit = 1
-$options.abort_when_failures_exceed = Float::INFINITY
-
-# start supporting some basic command line filtering of which browsers we run against
-opt_parser = OptionParser.new do |opts|
-  opts.banner = "Usage: runner.rb [options] \
-    Example: runner.rb -b chrome -o 7 -v 31 -f features/sharepage.feature \
-    Example: runner.rb -d localhost:3000 -t \
-    Example: runner.rb -l \
-    Example: runner.rb -r"
-  opts.separator ""
-  opts.separator "Specific options:"
-  opts.on("-c", "--config BrowserConfigName,BrowserConfigName1", Array, "Specify the name of one or more of the configs from ") do |c|
-    $options.config = c
-  end
-  opts.on("-b", "--browser BrowserName", String, "Specify a browser") do |b|
-    $options.browser = b
-  end
-  opts.on("-o", "--os_version OS Version", String, "Specify an os version") do |os|
-    $options.os_version = os
-  end
-  opts.on("-v", "--browser_version Browser Version", String, "Specify a browser version") do |bv|
-    $options.browser_version = bv
-  end
-  opts.on("-f", "--feature Feature", Array, "Single feature or comma separated list of features to run") do |f|
-    $options.feature = f
+  start_time = Time.now
+  ENV['BATCH_NAME'] = "#{GIT_BRANCH} | #{start_time}"
+
+  open_log_files
+  configure_for_eyes if eyes?
+  report_tests_starting
+  generate_status_page(start_time) if options.with_status_page
+
+  run_results = Parallel.map(browser_feature_generator, parallel_config(options.parallel_limit)) do |browser, feature|
+    run_feature browser, feature, options
   end
-  opts.on("-l", "--local", "Use local webdriver (not Saucelabs) and local domains") do
-    $options.local = 'true'
-    $options.pegasus_domain = 'localhost.code.org:3000'
-    $options.dashboard_domain = 'localhost.studio.code.org:3000'
-    $options.hourofcode_domain = 'localhost.hourofcode.com:3000'
+
+  # Produce a final report if we aborted due to excess failures
+  if $failed_features > options.abort_when_failures_exceed
+    abandoned_message = "Test run abandoned; limit of #{options.abort_when_failures_exceed} failed features was exceeded."
+    ChatClient.log abandoned_message, color: 'red'
   end
-  opts.on("-p", "--pegasus Domain", String, "Specify an override domain for code.org, e.g. localhost.code.org:3000") do |p|
-    if p == 'localhost:3000'
-      print "WARNING: Some tests may fail using '-p localhost:3000' because cookies will not be available.\n"\
-            "Try '-p localhost.code.org:3000' instead (this is the default when using '-l').\n"
+
+  # If we aborted for some reason we may have no run results, and should
+  # exit with a failure code.
+  return 1 if run_results.nil?
+
+  report_tests_finished start_time, run_results
+  run_results.count {|succeeded, _, _| !succeeded}
```

Comment:
I'm not changing existing behavior here.  As far as I know nothing is consuming this status code beyond its simple success (zero)/failure (nonzero) state, but I haven't investigated in depth. Do you think there's something more useful we could do with these status codes?  I'm open to the discussion, but would prefer to add features in a new PR since this one is already monstrous.